feat(parsers.avro): Add Apache Avro parser plugin

athornton opened the pull request:
# Required for all PRs

<!-- Before opening a pull request you should run the following checks to make sure the CI will pass.

make check
make check-deps
make test
make docs

-->

<!-- Complete the tasks in the following list. Change [ ] to [x] to
show completion. -->

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

<!-- Link to issues that describe the need for the change. Issues
should include context that will help reviewers understand why the
change is needed.

Make sure to link issues and using a keyword like "resolves #1234".
https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

resolves #1630

This is a replacement for https://github.com/influxdata/telegraf/pull/7732 since the original author (@emanuele-falzone ) has gone silent.

<!-- Finally, include a summary of the code change itself. This
description should tell reviewers how the issues were resolved.

example: Fixed an off by one error in counter variable in type FooBar.

example: Added an input plugin to gather yak shaving metrics using
golang library yaktech/shaver. -->

This builds on Emanuele Falzone's work to allow ingestion from Avro serialized format.  It can either connect to a schema registry or a schema can be specified in the parser.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+	return nil
+}
+
+func (p *Parser) InitFromConfig(config *parsers.Config) error {
+	p.MetricName = config.MetricName
+	p.SchemaRegistry = config.AvroSchemaRegistry
+	p.Schema = config.AvroSchema
+	p.Measurement = config.AvroMeasurement
+	p.Tags = config.AvroTags
+	p.Fields = config.AvroFields
+	p.Timestamp = config.AvroTimestamp
+	p.TimestampFormat = config.AvroTimestampFormat
+	p.FieldSeparator = config.AvroFieldSeparator
+	p.DefaultTags = config.DefaultTags
+
+	return p.Init()
+}
```
This is not needed for new parsers. Please remove it.

athornton replied:
I did, but my primary sticking point right now is that I do not understand how to skip the compatibility test, so right now my test build is failing.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+
+func (p *Parser) Init() error {
+	p.TimeFunc = time.Now
+	if p.Schema == "" && p.SchemaRegistry == "" {
+		err := fmt.Errorf("one of SchemaRegistry or Schema must be specified")
+		return err
+	}
```
Please use the config options here
```suggestion
	if p.Schema == "" && p.SchemaRegistry == "" {
		return errors.New("one of 'schema_registry' or 'schema' must be specified")
	}
```

Furthermore, I think having both specified also should bring up an error shouldn't it?

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+	DiscardArrays   bool     `toml:"avro_discard_arrays"`
+	FieldSeparator  string   `toml:"avro_field_separator"`
+	DefaultTags     map[string]string
+	TimeFunc        func() time.Time
```
Why do you need to manipulate time here?

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+	TimestampFormat string   `toml:"avro_timestamp_format"`
+	DiscardArrays   bool     `toml:"avro_discard_arrays"`
+	FieldSeparator  string   `toml:"avro_field_separator"`
+	DefaultTags     map[string]string
```
Please change to
```suggestion
	DefaultTags     map[string]string `toml:"-"`
```
to prevent setting this through the configuration.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var schema string
+
+	if len(buf) < 5 {
+		err := fmt.Errorf("buf is %d bytes; must be at least 5", len(buf))
+		return nil, err
+	}
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
```
Shoudn't the underlying parser library deal with this?

athornton replied:
If you're doing the usual use case, where you have a schema registry and the binary data corresponding to the packed encoding of that schema, sure.  But in the parent PR, there was discussion from someone who wanted to manually specify the schema rather than retrieve it from the registry.  Apparently not all Avro users are also Kafka users.

See https://github.com/influxdata/telegraf/pull/7732#discussion_r463325701 .  Maybe @rodrigobrochado is still around and can comment?  Anyway, I put in the guard so we had a slightly more controlled error if somehow we got passed a short set of bytes (rather than taking a slice that wasn't there), which was @rodrigobrochado's suggestion on the original PR.

In the not-Kafka case, I don't actually know if I should just decode the whole object, rather than just the sixth and subsequent bytes, because it's not my use case.  I can change to that if you think it's a good idea.

srebhan replied:
@athornton and @rodrigobrochado thanks for clarifying! Can you please make this a small helper function e.g. `extractScheme(buf []byte) (int, []byte, error)` and explain what you are doing with a small comment?

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
+
+	switch {
+	case p.SchemaRegistry != "":
+		schemaRegistry := NewSchemaRegistry(p.SchemaRegistry)
+		retrSchema, err := schemaRegistry.getSchema(schemaID)
+		if err != nil {
+			return nil, err
+		}
+		schema = retrSchema
+	default:
+		schema = p.Schema
+	}
```
Why not use a simple `if` statement? Furthermore, do we really need to construct the registry for each parsing run or can this be done in `Init()`?

athornton replied:
It can certainly be an if statement.  We do, however, need to look up each schema, and each record may have a different schema--my use case consumes data from many, often dynamically-updated (so therefore the schema ID changes), schemas.  However, getSchema memoizes its results, so we only look up a given schema ID from the registry the first time we see that ID.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	codec, err := goavro.NewCodec(schema)
+
```
Please remove the blank line here.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+func (p *Parser) parseTimestamp(timestamp interface{}) (time.Time, error) {
+	if timestamp == nil {
+		return p.TimeFunc(), nil
+	}
```
I would rather have this check on the caller side...

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+		return p.TimeFunc(), nil
+	}
+
+	if p.TimestampFormat == "" {
+		return p.TimeFunc(), fmt.Errorf("must specify timestamp format")
+	}
```
This should be checked in `Init()`.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+		return p.TimeFunc(), fmt.Errorf("must specify timestamp format")
+	}
+
+	metricTime, err := internal.ParseTimestamp(p.TimestampFormat, timestamp, "UTC")
+	if err != nil {
+		return p.TimeFunc(), err
+	}
+
+	return metricTime, nil
```
```suggestion
	return := internal.ParseTimestamp(p.TimestampFormat, timestamp, "UTC")
```

athornton replied:
These together argue for the elimination of parseTimestamp() from the parser, which is fine.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+	return metricTime, nil
+}
+
+func (p *Parser) createMetric(native interface{}, schema string) (telegraf.Metric, error) {
```
Is there always only one metric in the data or is it possible to send "arrays" of metrics?

athornton replied:
One metric per message, at least according to the README.md I inherited (and I have not seen any behavior to suggest otherwise).

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+
+	for _, tag := range p.Tags {
+		if value, ok := deep[tag]; ok {
+			tags[tag] = fmt.Sprintf("%v", nestedValue(value))
```
How about [`internal.ToString()`](https://github.com/influxdata/telegraf/blob/f803c6b410036ebaaf1dfbba175ad898c38ab50a/internal/type_conversions.go#L8)?

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+	for _, tag := range p.Tags {
+		if value, ok := deep[tag]; ok {
+			tags[tag] = fmt.Sprintf("%v", nestedValue(value))
+		} else {
+			p.Log.Infof("tag %s value was %v; not added to tags", tag, value)
+		}
```
This rather sounds like a debug message. Otherwise you may risk log-spamming...

athornton replied:
More likely I should make it a warning.  If we fail to unpack the value, something weird is going on.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+			p.Log.Infof("tag %s value was %v; not added to tags", tag, value)
+		}
+	}
+	fieldList := make([]string, len(p.Fields), (cap(p.Fields)+1)*2)
+	copy(fieldList, p.Fields)
+	if len(fieldList) == 0 { // Get fields from whatever we just unpacked
+		for k := range deep {
+			if _, ok := tags[k]; !ok {
+				fieldList = append(fieldList, k)
+			}
+		}
+	}
+	for _, field := range fieldList {
+		if value, ok := deep[field]; ok {
+			fields[field] = nestedValue(value)
+		} else {
+			p.Log.Infof("field %s value was %v; not added to fields", field, value)
+		}
+	}
```
Why is this different to the `tags` part?

athornton replied:
If Fields is not explicitly specified, we retrieve the fields from the schema associated with the binary Avro data.

Avro doesn't have a concept of Tag vs Field, so if we have any tags, they will necessarily be explicitly specified, but fields can be dynamically discovered.  And while tags are necessarily strings, fields don't have to be.

I'm oversizing the field list capacity in anticipation that I'm going to need to unpack arrays, because that's what a lot of my own data looks like.  I'm OK with leaving the capacity out and just eating the (common-for-me, but maybe not everyone) hit of resizing the field list when I unpack the arrays I get.

srebhan replied:
Again, a small comment in the code would help the reader... :-)

athornton replied:
Done.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+			p.Log.Infof("field %s value was %v; not added to fields", field, value)
+		}
+	}
+	var schemaObj map[string]interface{}
+
+	err = json.Unmarshal([]byte(schema), &schemaObj)
+	if err != nil {
+		return nil, err
+	}
+	if len(fields) == 0 {
+		return nil, fmt.Errorf("number of fields is 0; unable to create metric")
+	}
+	name := ""
+	if p.Measurement != "" {
+		name = p.Measurement
+	} else {
+		// get Measurement name from schema
+		nsStr, ok := schemaObj["namespace"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine namespace from schema %s", schema)
+		}
+		nStr, ok := schemaObj["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine name from schema %s", schema)
+		}
+		name = nsStr + "." + nStr
+	}
+	if name == "" {
+		name = p.MetricName
+	}
+	if name == "" {
+		return nil, fmt.Errorf("could not determine measurement name")
+	}
```
I'm pretty sure this can be done in `Init()`.

athornton replied:
It can't, because the schema can (and does) change between messages.  So if we haven't explicitly set the Measurement name, we have to unpack it from the schema (as `<namespace>.<name>`) .  If somehow that fails (which it should not for a correctly-formed schema), then we use MetricName as a fallback, and if we don't have even that, then we have no idea what the metric name should be.

This logic does, however, need a bit of rework, because the namespace is allowed to be empty (https://avro.apache.org/docs/1.11.1/specification/)

srebhan replied:
Ok I see, please also add a small comment to help the reader.

athornton replied:
Done.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,238 @@
+		// Any non-scalars end up being a nil field
+		return metric.New(name, tags, fields, metricTime), nil
+	}
+	// But if we do it this way, we flatten any compound structures,
+	// including arrays.  Goavro is only going to hand us back
+	// arrays, not maps.
+	// The default (the separator string is empty) is equivalent to
+	// what streamreactor does.
+	sep := flatten.SeparatorStyle{
+		Before: "",
+		Middle: p.FieldSeparator,
+		After:  "",
+	}
+	flat, err := flatten.Flatten(fields, "", sep)
+	if err != nil {
+		return nil, err
+	}
+	return metric.New(name, tags, flat, metricTime), nil
```
Shouldn't we have two distinct function switched at `Init()` time instead of first fill the fields and then throw them away?

athornton replied:
We don't necessarily know ahead of time if a field is going to be an array or not (because each record we get is a tuple of schema and data, and the schema is often nothing we have met before).

If you just throw an array field at telegraf/metric it doesn't know what to do with it so you just don't get that field in the output metric.  This *might* be what a user wants, but it's certainly not what my use case requires: I want the arrays flattened into multiple columns.  Since I know that *eventually* I want the column names with an underscore separator, but right now (to maintain compatibility with our streamreactor implementation while we transition from it + InfluxDBv1 to the avro parser + InfluxDBv2) I do not, so I made that a configurable choice.

Either way we're calling metric.New() ; and while I guess we could examine each of our fields and determine whether metric.New() is going to drop that field (because it's an array) and pre-discard it, there doesn't seem to me to be a lot of point in doing so.

We could also just make the decision that if you get an array field of course you want it flattened into multiple columns.  That's what streamreactor does, and it's what I want, and I suspect that it is the overwhelmingly-more-common use case, but I'm not sure we want to remove the flatten-or-discard choice from the user entirely.

srebhan replied:
Understood. Seems like I overlooked the dynamic of this thing. :-) Anyway, please split this function into an "for scalars" and "for arrays" function and call the relevant one. For me it is confusing if first fields are set and later be overwritten without any clear indication of what's going on.

athornton replied:
Done.

srebhan commented on plugins/parsers/avro/parser_test.go:
```diff
@@ -0,0 +1,269 @@
+	return time.Unix(3600, 0)
+}
+
+func TestBasicAvroMessage(t *testing.T) {
```
Can you maybe convert your test-cases to the scheme used in [json_v2](https://github.com/influxdata/telegraf/tree/master/plugins/parsers/json_v2/testdata) or (partially) in [xpath](https://github.com/influxdata/telegraf/tree/master/plugins/parsers/xpath/testcases)? This would ease adding more test-cases and give people a direct configuration example to play with.

athornton replied:
I'll do that as a separate commit, in that that's going to take understanding more of the tests than I have so far (that's all inherited from the previous PR).

srebhan replied:
Ok, but please take a look as this allows for fast addition of new test-cases by more or less only copying in the user's config and input data.

athornton replied:
I have done an initial implementation of this.  The rest of the stuff you asked for is much easier.

srebhan commented on config/config.go:
```diff
@@ -1217,6 +1217,7 @@ func (c *Config) missingTomlField(_ reflect.Type, key string) error {
 
 	// Serializer options to ignore
 	case "prefix", "template", "templates",
+		"avro_schema", "avro_schema_registry", "avro_measurement", "avro_tags", "avro_fields", "avro_timestamp", "avro_timestamp_format", "avro_field_separator",
```
Can you please check if the parser works without setting those as they are scheduled for removal anyway. Furthermore, serializers would be the wrong place anyway. ;-)

srebhan commented on plugins/parsers/registry.go:
```diff
@@ -160,6 +160,16 @@ type Config struct {
 	CSVMetadataSeparators []string `toml:"csv_metadata_separators"`
 	CSVMetadataTrimSet    string   `toml:"csv_metadata_trim_set"`
 
+	// Avro configuration
+	AvroSchemaRegistry  string   `toml:"avro_schema_registry"`
+	AvroSchema          string   `toml:"avro_schema"`
+	AvroMeasurement     string   `toml:"avro_measurement"`
+	AvroTags            []string `toml:"avro_tags"`
+	AvroFields          []string `toml:"avro_fields"`
+	AvroTimestamp       string   `toml:"avro_timestamp"`
+	AvroTimestampFormat string   `toml:"avro_timestamp_format"`
+	AvroFieldSeparator  string   `toml:"avro_field_separator"`
+
```
I don't think this is necessary to add this here.

srebhan commented on plugins/parsers/registry_test.go:
```diff
@@ -16,11 +16,12 @@ import (
-		XPathProtobufFile: "xpath/testcases/protos/addressbook.proto",
-		XPathProtobufType: "addressbook.AddressBook",
-		JSONStrict:        true,
+		MetricName:         "parser_compatibility_test",
+		CSVHeaderRowCount:  42,
+		XPathProtobufFile:  "xpath/testcases/protos/addressbook.proto",
+		XPathProtobufType:  "addressbook.AddressBook",
+		JSONStrict:         true,
+		AvroSchemaRegistry: "https://localhost:8081",
```
Please skip backward compatibility test for `avro` as it doesn't need to be backward compatible.

srebhan reviewed (commented):
Thanks @athornton for reviving this parser! I have some comments, nothing too big. The only part that concerns me is the addition of time-formats. Please avoid adding those and rather add a `round_timestamps_to` option in your parser, avoiding the combinatorics of formats and rounding.

srebhan commented:
@athornton please also rebase to latest master to get CircleCI back functional.

athornton commented:
Thank you for the detailed review.  I'll get to work on it.  I have no objection to a `round_timestamps_to` config item rather than my initial implementation (the reason to round at all basically comes down to https://docs.influxdata.com/influxdb/v2.4/reference/faq/#does-the-precision-of-the-timestamp-matter , and I at least find it much easier to eyeball the data if all digits past the precision we care about are zero rather than the deterministic-but-basically-random stuff that the conversion gives us).

athornton commented:
Also take note that a lot of the motivation for what I'm doing here really isn't obvious until https://github.com/influxdata/telegraf/pull/11831 .  Basically, this is the avro parsing piece that will then make it possible to extend the Kafka input consumer with dynamic regexp-based topic detection and periodic scanning for new topics.  These are features we need (and that streamreactor gives us but only for InfluxDBv1).

athornton commented:
@srebhan I think I'm a little stuck.  I've rebased against current master and CircleCI is still not working, and I don't know how to tell test_registry.go to *not* invoke the backwards-compatibility check.

```
adam@ixitxachitl:~/git/telegraf/plugins/parsers$ go test
--- FAIL: TestRegistry_BackwardCompatibility (0.01s)
    registry_test.go:47: testing "csv"...
    registry_test.go:47: testing "graphite"...
    registry_test.go:47: testing "grok"...
    registry_test.go:47: testing "nagios"...
    registry_test.go:47: testing "value"...
    registry_test.go:47: testing "xpath_json"...
    registry_test.go:47: testing "influx_upstream"...
    registry_test.go:47: testing "json_v2"...
    registry_test.go:47: testing "prometheusremotewrite"...
    registry_test.go:47: testing "wavefront"...
    registry_test.go:47: testing "avro"...
    registry_test.go:60:
        	Error Trace:	/Users/adam/git/telegraf/plugins/parsers/registry_test.go:60
        	Error:      	Received unexpected error:
        	            	exactly one of 'schema_registry' or 'schema' must be specified
        	Test:       	TestRegistry_BackwardCompatibility
FAIL
exit status 1
```

athornton commented on internal/internal.go:
```diff
@@ -276,14 +287,20 @@ func parseUnix(format string, timestamp interface{}) (time.Time, error) {
+		return time.Unix(integer,
+			int64(float64(fractional)/float64(1e3)+0.5)*1e3).UTC(), nil
+	case "unix", "unix_float_ns":
+		return time.Unix(integer, fractional).UTC(), nil
```
OK, I have replaced this implementation with `round_timestamp_to` as a parser option.

athornton commented:
My suspicion is that the CircleCI config isn't working because this is the lsst-sqre fork of Telegraf, so we really don't have CircleCI permissions.  The configuration files are identical.

athornton commented:
I added a very crude way to skip parsers that are not backward-compatible: a map at the top of the test file, and a check to see if the parser's name was in that map.  Surely there's a better way of doing this, but I didn't find it.  Happy to change it if someone points me in the right direction.

Other than that...I think I've addressed all the commentary except improving the test suite.

srebhan commented on plugins/parsers/registry_test.go:
```diff
@@ -44,6 +49,10 @@ func TestRegistry_BackwardCompatibility(t *testing.T) {
 	}
 
 	for name, creator := range parsers.Parsers {
+		if _, ok := notBackwardsCompatible[name]; ok {
+			t.Logf("skipping %q...", name)
+			continue
+		}
```
Just use
```suggestion
		if name == "avro" {
			t.Logf("skipping %q...", name)
			continue
		}
```
and we can switch to a `switch` statement if more are added. This is a temporary thing anyway and backward compatibility will be removed with 2.0 anyway.

srebhan commented:
@athornton regarding timestamps, please also take a look at #11875! While not directly obvious, this changes the timestamp parsing to handle timestamps by keeping the precision to nanoseconds. If you need more, we should try to talk about this in a separate issue describing your expected behavior.

athornton commented:
Oh, that timestamp work looks very nice.  I'll try to get to it soon, and I will change the skip-compatibility test.

The test suite is proving difficult for me--basically, without an InitFromConfig it's not clear to me how to *load* a new telegraf.conf for a parsing run and initialize the parser from it (Init() fails, obviously, because it wants schema or schemaURL, so I have to construct the Parser with fields set from the config so that Init() will work; but that means I need to extract them from the config and there's some unmarshaling I am missing) .  I think I am about to end up basically manually doing what InitFromConfig did in my test suite.

athornton commented:
I've added a file-based test suite, and use it to check for all the Init() errors (and a happy-path case).  It could use some more attention, but I think this begins to get the point across.

athornton commented:
@srebhan I've grabbed the changes from #11875 and applied them to this one.  We've tested it for our use case and it works just fine, so I dropped our round_timestamp_to stuff.

athornton commented:
@srebhan is there anything else you need me to do on this one?

athornton commented:
@srebhan repinging: I think I've done what I can here (the CircleCI seems to legit be "I do not have permission to do so").  Is there anything else I need to do on this before it's ready to merge?

powersj commented:
I'm going to close and re-open this PR to see if it can get circleci going again

powersj commented:
> will only run from a personal fork, not an organizational fork?

We have had some folks from Home Depot submit from their org, so I am not sure, but it did run yesterday:

https://app.circleci.com/pipelines/github/influxdata/telegraf/13402/workflows/e22fa15c-e1db-4448-b1ae-3dd54abf7664

I'm not sure why there is still a "CircleCI Pipeline" entry in the list of checks, but the other checks were from yesterday

athornton commented:
I'd really love to be able to land this so I can shift over to the (much smaller) add regexp support to kafka topic ingestion PR that sits on top of this.  Is there anything I can do at this point?  The Home Depot data seems pretty conclusive that it's not a personal/org fork problem.

powersj commented:
> Is there anything I can do at this point? 

Looks like the tests didn't even run as there are go.mod issues. Running `go mod tidy` and pushing the little change should get you a clean run. I can watch to see if tests run again as well.

After that I can ping @srebhan next week as it looked like you were close to wrapping this up.

athornton commented:
OK.  Looks like we're back to everything but CircleCI passing.

srebhan reviewed (commented):
Thanks for the update @athornton and sorry for the late reply! I have some more minor comments but we are getting closer to be able to merge this!

srebhan commented:
@athornton any update on this PR?

srebhan commented:
@athornton any way you finish this PR?

athornton commented:
Yeah, sorry, I got pulled off to do other things and didn't even see the November commentary.  I will get back to it and we can get this thing over the finish line.

athornton commented on go.mod:
```diff
@@ -103,13 +103,15 @@ require (
 	github.com/karrick/godirwalk v1.17.0
 	github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
 	github.com/kolo/xmlrpc v0.0.0-20201022064351-38db28db192b
-	github.com/lxc/lxd v0.0.0-20220920163450-e9b4b514106a
+	github.com/linkedin/goavro/v2 v2.12.0
+	github.com/lxc/lxd v0.0.0-20220809104211-1aaea4d7159b
```
Doubtful.  I'll regenerate all the dependencies and hopefully this will go away.

srebhan commented:
@athornton THANKS! Please let me know if this is ready for another review round!

athornton commented:
Not ready yet.  I merged the suggestions that I was sure were uncontroversial, but some of the other stuff I'm going to need to see if I had a reason for doing it the way I did.  I'd hoped to work on it today, but I got a high-priority interrupt.  Hopefully tomorrow I get a little while to work on it.

athornton commented:
I have a question.  I had planned to do https://github.com/influxdata/telegraf/pull/11831 as a separate PR on top of this.  However, I suspect that our use case, where we get Avro schemas (and update them dynamically) from a kafka broker is actually a very common one.  Since I'm digging back into this one, rolling that one in might be a good idea, or I could try to land that one separately after this one is merged.  Which do you think is a better idea?

powersj commented:
> I could try to land that one separately after this one is merged

I would much rather see two PRs over one very large one.

athornton commented:
Still working on it.  Got it passing its old unit tests again and now I'm on implementing the requested changes to the test suite, but I have had other stuff that needs attention today.

athornton commented:
@srebhan I failed to understand how to use the file plugin and Gather to eliminate a bunch of the setup.

The CircleCI Linux failures, I think, may be because I am not running Golang 1.20 yet on my Mac (using Brew, golang is still at 1.19).

At any rate, the unit tests are again passing, as is make check/make test.  If you wanted to just fix the test case to do the configuration correctly I'd appreciate it; I'm kind of swamped in my day job again.  If instead of doing that, you would point me to docs on what it is I'm supposed to do with Gather I'll be happy to take another crack at it, but it might be a little while.

powersj commented:
@athornton here is a diff to clean up the `go mod tidy` issues:

```diff
diff --git a/go.mod b/go.mod
index 3b95a4eca..1f992bfc8 100644
--- a/go.mod
+++ b/go.mod
@@ -107,7 +107,7 @@ require (
        github.com/jackc/pgtype v1.12.0
        github.com/jackc/pgx/v4 v4.17.1
        github.com/james4k/rcon v0.0.0-20120923215419-8fbb8268b60a
-       github.com/jeremywohl/flatten v1.0.1
+       github.com/jeremywohl/flatten/v2 v2.0.0-20211013061545-07e4a09fb8e4
        github.com/jhump/protoreflect v1.8.3-0.20210616212123-6cc1efa697ca
        github.com/jmespath/go-jmespath v0.4.0
        github.com/kardianos/service v1.2.2
@@ -199,11 +199,6 @@ require (
        modernc.org/sqlite v1.19.2
 )
 
-require (
-       github.com/jeremywohl/flatten/v2 v2.0.0-20211013061545-07e4a09fb8e4 // indirect
-       github.com/pkg/xattr v0.4.9 // indirect
-)
-
 require (
        cloud.google.com/go v0.107.0 // indirect
        cloud.google.com/go/compute v1.14.0 // indirect
@@ -380,6 +375,7 @@ require (
        github.com/pion/udp v0.1.4 // indirect
        github.com/pkg/browser v0.0.0-20210911075715-681adbf594b8 // indirect
        github.com/pkg/sftp v1.13.5 // indirect
+       github.com/pkg/xattr v0.4.9 // indirect
        github.com/pmezard/go-difflib v1.0.0 // indirect
        github.com/power-devops/perfstat v0.0.0-20220216144756-c35f1ee13d7c // indirect
        github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475 // indirect
diff --git a/go.sum b/go.sum
index 0a3b55104..585339e8c 100644
--- a/go.sum
+++ b/go.sum
@@ -1378,8 +1378,6 @@ github.com/jcmturner/gokrb5/v8 v8.4.3 h1:iTonLeSJOn7MVUtyMT+arAn5AKAPrkilzhGw8wE
 github.com/jcmturner/gokrb5/v8 v8.4.3/go.mod h1:dqRwJGXznQrzw6cWmyo6kH+E7jksEQG/CyVWsJEsJO0=
 github.com/jcmturner/rpc/v2 v2.0.3 h1:7FXXj8Ti1IaVFpSAziCZWNzbNuZmnvw/i6CqLNdWfZY=
 github.com/jcmturner/rpc/v2 v2.0.3/go.mod h1:VUJYCIDm3PVOEHw8sgt091/20OJjskO/YJki3ELg/Hc=
-github.com/jeremywohl/flatten v1.0.1 h1:LrsxmB3hfwJuE+ptGOijix1PIfOoKLJ3Uee/mzbgtrs=
-github.com/jeremywohl/flatten v1.0.1/go.mod h1:4AmD/VxjWcI5SRB0n6szE2A6s2fsNHDLO0nAlMHgfLQ=
 github.com/jeremywohl/flatten/v2 v2.0.0-20211013061545-07e4a09fb8e4 h1:eA9wi6ZzpIRobvXkn/S2Lyw1hr2pc71zxzOPl7Xjs4w=
 github.com/jeremywohl/flatten/v2 v2.0.0-20211013061545-07e4a09fb8e4/go.mod h1:s9g9Dfls+aEgucKXKW+i8MRZuLXT2MrD/WjYpMnWfOw=
 github.com/jessevdk/go-flags v1.4.0/go.mod h1:4FA24M0QyGHXBuZZK/XkWh8h0e1EYbRYJSGM75WSRxI=
@@ -1531,7 +1529,6 @@ github.com/lightstep/lightstep-tracer-common/golang/gogo v0.0.0-20190605223551-b
 github.com/lightstep/lightstep-tracer-go v0.18.1/go.mod h1:jlF1pusYV4pidLvZ+XD0UBX0ZE6WURAspgAczcDHrL4=
 github.com/linkedin/goavro/v2 v2.12.0 h1:rIQQSj8jdAUlKQh6DttK8wCRv4t4QO09g1C4aBWXslg=
 github.com/linkedin/goavro/v2 v2.12.0/go.mod h1:KXx+erlq+RPlGSPmLF7xGo6SAbh8sCQ53x064+ioxhk=
-github.com/linuxkit/virtsock v0.0.0-20201010232012-f8cee7dfc7a3/go.mod h1:3r6x7q95whyfWQpmGZTu3gk3v2YkMi05HEzl7Tf7YEo=
 github.com/logrusorgru/aurora v0.0.0-20181002194514-a7b3b318ed4e/go.mod h1:7rIyQOR62GCctdiQpZ/zOJlFyk6y+94wXzv6RNZgaR4=
 github.com/logzio/azure-monitor-metrics-receiver v1.0.0 h1:TAzhIZL2ueyyc81qIw8FGg4nUbts4Hvc3oOxSobY1IA=
 github.com/logzio/azure-monitor-metrics-receiver v1.0.0/go.mod h1:UIaQ7UgxZ8jO3L0JB2hctsHFBbZqL6mbxYscQAeFpl4=
```

athornton commented:
The patch didn't apply cleanly for me (I copied the patch, saved as a file, and patched), so I applied it by hand.  I think I got it right but I have the same two failures.

athornton commented:
But then I decided to just specify go 1.19 in go.mod so I could run go mod tidy locally.

srebhan commented on plugins/parsers/avro/parser_test.go:
```diff
@@ -0,0 +1,188 @@
+	return &p, nil
+}
+
+func TestMultipleConfigs(t *testing.T) {
```
I think your implementation of the tests is too complicated. Check the [binary parser test](https://github.com/influxdata/telegraf/blob/d1d9737da6f1dabd3350e5fdd80f7003ed207a81/plugins/parsers/binary/parser_test.go#L1407) for an example.
Basically there should be _no need at all_ for the `buildParser` function, nor for defining the input file. All of those values should be defined in the `telegraf.conf` file(s)!

srebhan commented on plugins/parsers/avro/schema_registry.go:
```diff
@@ -0,0 +1,61 @@
+package avro
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/linkedin/goavro/v2"
+	"net/http"
```
Please group imports
```suggestion
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/linkedin/goavro/v2"
```

srebhan commented on plugins/parsers/avro/schema_registry.go:
```diff
@@ -0,0 +1,61 @@
+	"net/http"
+)
+
+type SchemaAndCodec struct {
+	Schema string
+	Codec  *goavro.Codec
+}
+
+type SchemaRegistry struct {
```
Do those structs need to be exported?

srebhan commented on plugins/parsers/avro/schema_registry.go:
```diff
@@ -0,0 +1,61 @@
+	cache map[int]*SchemaAndCodec
+}
+
+const (
+	schemaByID = "%s/schemas/ids/%d"
+)
```
```suggestion
const schemaByID = "%s/schemas/ids/%d"
```

srebhan commented on plugins/parsers/avro/schema_registry.go:
```diff
@@ -0,0 +1,61 @@
+	schemaByID = "%s/schemas/ids/%d"
+)
+
+func NewSchemaRegistry(url string) *SchemaRegistry {
```
Same here, does the function needs to be exported?

srebhan commented on plugins/parsers/avro/schema_registry.go:
```diff
@@ -0,0 +1,61 @@
+
+	var jsonResponse map[string]interface{}
+
+	err = json.NewDecoder(resp.Body).Decode(&jsonResponse)
+	if err != nil {
+		return nil, err
+	}
```
This will make use of a limited-scope `err` variable
```suggestion
	if err := json.NewDecoder(resp.Body).Decode(&jsonResponse); err != nil {
		return nil, err
	}
```

srebhan commented on plugins/parsers/avro/schema_registry.go:
```diff
@@ -0,0 +1,61 @@
+		return nil, err
+	}
+	retval := &SchemaAndCodec{Schema: schemaValue, Codec: codec}
+	sr.cache[id] = retval
```
I guess you are sure that there is no concurrent access to the `sr.cache` map!? Otherwise this needs to be protected by a mutex.

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,218 @@
+	}
+	// Nothing?  Give up.
+	if name == "" {
+		return nil, fmt.Errorf("could not determine measurement name")
```
```suggestion
		return nil, errors.New("could not determine measurement name")
```

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,218 @@
+
+func (p *Parser) Init() error {
+	if (p.Schema == "" && p.SchemaRegistry == "") || (p.Schema != "" && p.SchemaRegistry != "") {
+		return fmt.Errorf("exactly one of 'schema_registry' or 'schema' must be specified")
```
```suggestion
		return errors.New("exactly one of 'schema_registry' or 'schema' must be specified")
```

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,218 @@
+		p.registryObj = NewSchemaRegistry(p.SchemaRegistry)
+	}
+	if p.TimestampFormat == "" {
+		return fmt.Errorf("must specify 'timestamp_format'")
```
```suggestion
		return errors.New("must specify 'timestamp_format'")
```

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,218 @@
+		candidate[fld] = data[fld] // 1-item map
+		flat, err := flatten.Flatten(candidate, "", sep)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to flatten field %s: %v", fld, err)
```
Error messages need to start lower-case. Furthermore, please always wrap errors so we can check for the error-type later on
```suggestion
			return nil, fmt.Errorf("flatten field %q failed: %w", fld, err)
```

srebhan commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	if len(metrics) != 1 {
+		return nil, fmt.Errorf("line contains multiple metrics")
```
In case your error does not have any parameter use `errors.New` instead of `fmt.Errorf`
```suggestion
		return nil, errors-New("line contains multiple metrics")
```

Please also check other locations in your code and correct if applicable.

athornton commented:
I will get to these comments and suggestions when I am able.  Thank you, and thank you especially for pointing me to the binary parser as a model for what I should be doing.

athornton commented:
At this point I think I've got everything but the test rewrite handled.

athornton commented:
@srebhan :

OK, I think I see conceptually what you're saying: all the convenience tools where I take the JSON representations of the schema and the message, and then call jsonToAvroMessage to generate the Avro format input, should be replaced by a binary input message (the output of jsonToAvroMessage) and a simple test of whether that works?  Although I could put the schema or even both the schema and the message into telegraf.conf, in actual use the schema will be externally-given (almost always, it will come from a schema registry), and obviously the message is coming in over the wire.

So it feels like we want a much simpler test, of "Avro format binary data" as the input...but if we want more test cases at some point, I don't want to throw away the tooling to create those messages, because in practice, generating the test data will be done by matching a schema and message and generating the Avro data from them, rather than generating the wire protocol by hand.  Where should that tooling go?

srebhan commented:
> OK, I think I see conceptually what you're saying: all the convenience tools where I take the JSON representations of the schema and the message, and then call jsonToAvroMessage to generate the Avro format input, should be replaced by a binary input message (the output of jsonToAvroMessage) and a simple test of whether that works? Although I could put the schema or even both the schema and the message into telegraf.conf, in actual use the schema will be externally-given (almost always, it will come from a schema registry), and obviously the message is coming in over the wire.

Exactly. Put the binary messages there and let the file input read them.

> So it feels like we want a much simpler test, of "Avro format binary data" as the input...but if we want more test cases at some point, I don't want to throw away the tooling to create those messages, because in practice, generating the test data will be done by matching a schema and message and generating the Avro data from them, rather than generating the wire protocol by hand. Where should that tooling go?

I don't think that tool should be in Telegraf. It's not Telegraf's task to create those messages. If we add further test-cases it will likely be based on bug-reports, so it would be nice if the parser could print the binary message it receives and maybe even the schema as debug messages on error. We have added this for a few other plugins, e.g. GNMI one to be able to reproduce problems in tests...

athornton commented:
Hmm.  It's not quite that simple: messages may arrive as raw Avro binary data, as Avro single-object-encoding data, or as Confluent wire format.  So the parser will work on binary data, and if a parser registry is specified it will expect Confluent format.  So no explanatory comment yet.

However, I think I have the test suite rewritten now.

srebhan commented on plugins/parsers/avro/parser_test.go:
```diff
@@ -0,0 +1,86 @@
+			// Configure the plugin
+			cfg := config.NewConfig()
+			err := cfg.LoadConfig(configFilename)
+			if err != nil {
+				actualErrors = append(actualErrors, err.Error())
+			} else {
+				for _, input := range cfg.Inputs {
+					err := input.Init()
+					if err != nil {
+						actualErrors = append(actualErrors, err.Error())
+					} else {
+						err := input.Gather(&acc)
+						if err != nil {
+							actualErrors = append(actualErrors, err.Error())
+						}
+					}
+				}
+			}
```
I think we don't need to continue if loading the config fails or if initializing the parser fails. Test-cases should be for runtime errors only...
```suggestion
			require.NoError(t, err)

			for _, input := range cfg.Inputs {
				require.NoError(t, input.Init())

				if err := input.Gather(&acc); err != nil {
					actualErrors = append(actualErrors, err.Error())
				}
			}
```

srebhan commented on plugins/parsers/avro/schema_registry.go:
```diff
@@ -0,0 +1,59 @@
+	defer resp.Body.Close()
+
+	var jsonResponse map[string]interface{}
+
```
Maybe remove this empty line
```suggestion
```

srebhan reviewed (commented):
Awesome update @athornton! Just a few very small comments and then we are good to go I think...

athornton commented:
@srebhan I think we're ready.

srebhan reviewed (approved):
Thank you very much for driving this PR @athornton! Good job!

powersj commented on plugins/parsers/avro/README.md:
```diff
@@ -0,0 +1,97 @@
+`avro_timestamp_format` options together to set the time to a value in
+the parsed document.
+
+The `avro_timestamp` option specifies the field containing the time
+value.  It must be set to `unix`, `unix_ms`, `unix_us` or `unix_ns`.
```
The second sentence doesn't match the first right? Yes `avro_timestamp` is the field. The second sentence is referencing the `avro_timestamp_format` option, correct? And can't the format be a custom format?

powersj commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,220 @@
+	if (p.Schema == "" && p.SchemaRegistry == "") || (p.Schema != "" && p.SchemaRegistry != "") {
+		return errors.New("exactly one of 'schema_registry' or 'schema' must be specified")
+	}
+	if p.TimestampFormat == "" {
+		return errors.New("must specify 'timestamp_format'")
+	}
```
From the user side, why do they need always need to specify the format? If the timestamp is empty, then the current time is used and we know what to do. Shouldn't the format only be required if the timestamp is not nil?

powersj commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,220 @@
+			return nil, errors.New("first byte is not 0: not Confluent Wire Protocol")
+		}
+		schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+		schemastruct, err := p.registryObj.getSchemaAndCodec(schemaID)
+		if err != nil {
+			return nil, err
+		}
```
Should the call to `getSchemaAndCodec` 1) occur during init rather than at each gather cycle and 2) only occur once rather than every call to parse?

That has the potential to be a lot of HTTP calls.

powersj replied:
I see you marked this one resolved. Was there a previous discussions about keeping this here?

powersj commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,220 @@
+	if name == "" {
+		return nil, errors.New("could not determine measurement name")
+	}
+	return metric.New(name, tags, fields, time.Now()), nil
```
Isn't this always setting the time to now and never using the user's specified timestamp value?

powersj commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,220 @@
+	Timestamp       string            `toml:"avro_timestamp"`
+	TimestampFormat string            `toml:"avro_timestamp_format"`
+	FieldSeparator  string            `toml:"avro_field_separator"`
+	DefaultTags     map[string]string `toml:"-"`
```
Should this be the following, assuming you do not remove:

```suggestion
	DefaultTags     map[string]string `toml:"avro_default_tags"`
```

powersj commented on plugins/parsers/avro/README.md:
```diff
@@ -0,0 +1,97 @@
+  ## If this were set to "_", then it would be a_0="a", a_1="b".
+  # avro_field_separator = "_"
+
+  ## Default values for given tags: optional
+  # avro_default_tags = { "application": "hermes", "region": "central" }
```
Users can already specify tags on a per-plugin basis e.g. 

```toml
[[inputs.kafka_consumer]]
  tags = { "application": "hermes", "region": "central" }
```

Is this trying to accomplish the same thing? If so, I would prefer we stick to existing method.

powersj commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,220 @@
+	// Telegraf which items are our tags.
+	for _, tag := range p.Tags {
+		sTag, err := internal.ToString(data[tag])
+		if err != nil {
+			p.Log.Warnf("Could not convert %v to string for tag %q: %v", data[tag], tag, err)
+		}
+		tags[tag] = sTag
```
If we failed to convert, we shouldn't try to assign either (i.e. if err: warn + continue)

powersj reviewed (commented):
@athornton - huge thank you for your persistence on this PR. I have some questions in line.

powersj commented on plugins/parsers/avro/parser.go:
```diff
@@ -0,0 +1,245 @@
+	timestamp := time.Now()
+	if p.Timestamp != "" {
+		rawTime := fields[p.Timestamp]
+		timeIncs, ok := rawTime.(int64)
+		if !ok {
+			return nil, fmt.Errorf("timestamp %v could not be converted to int64", rawTime)
+		}
+		if p.TimestampFormat == "unix_us" {
+			timestamp = time.UnixMicro(timeIncs)
+		} else if p.TimestampFormat == "unix_ms" {
+			timestamp = time.UnixMilli(timeIncs)
+		} else if p.TimestampFormat == "unix_ns" {
+			// If it's a single Avro field, it's less than
+			// 2**63 - 1 ... which will last a while
+			timestamp = time.Unix(0, timeIncs)
+		} else {
+			timestamp = time.Unix(timeIncs, 0)
+		}
```
We have a helper to do this check out `timestamp, err := internal.ParseTimestamp(p.TimestampFormat, rawtime, "")`. I think it would be best to send in the timestamp as a string and you will need to import internal as well:

powersj commented:
@athornton thanks for the updates, I think we are down to two open questions:

1) The purpose of `DefaultTags` versus using the built-in method for defining tags for an input.
2) If `getSchemaAndCodec` shoudl be run on every `Parse`

Thanks!

athornton commented:
So, I think we may be done?  The schema lookup (when you have a schema registry) has to be done at each Parse(), but after the initial retrieval it's just a map lookup, so shouldn't be too costly.

powersj reviewed (approved):
Thanks for driving and contributing the new parser!
